src/Sylius/Bundle/ShopBundle/Resources/config/routing/cart.yml

```diff
@@ -26,17 +26,27 @@ sylius_shop_cart_clear:
         _sylius:
             redirect: sylius_shop_cart_summary
 
-sylius_shop_cart_item_add:
-    path: /add
-    methods: [POST]
-    defaults:
-        _controller: sylius.controller.cart_item:addAction
-        _sylius:
-            redirect: sylius_shop_cart_summary
-
 sylius_shop_cart_item_remove:
     path: /{id}/remove
     defaults:
         _controller: sylius.controller.cart_item:removeAction
         _sylius:
             redirect: sylius_shop_cart_summary
+
+sylius_shop_cart_item_add:
+    path: /add
+    methods: [GET, POST]
+    defaults:
+        _controller: sylius.controller.cart_item:addAction
+        _sylius:
+            factory:
+                method: createForProduct
+                arguments: [expr:service('sylius.repository.product').find($productId)]
+            form:
+                type: sylius_cart_item
+                options:
+                    product: expr:service('sylius.repository.product').find($productId)
```

Comment:
That's OK for this step of refactoring, later the cart item form should take the product from the data (already set on order item via factory), not from an option.